backend.native/tests/framework/values_generics/values.swift

```diff
@@ -0,0 +1,229 @@
+
+    let sd = SomeData(num: 33)
+    let nullColl = GenCollectionsNull<SomeData>(arg: sd, coll: [sd])
+    let nonNullColl = GenCollectionsNonNull<SomeData>(arg: sd, coll: [sd])
+
+    try assertEquals(actual: (nullColl.coll[0] as! SomeData).num, expected: 33)
+    try assertEquals(actual: nonNullColl.coll[0].num, expected: 33)
+    try assertEquals(actual: nonNullColl.arg.num, expected: 33)
+
+    let mixed = GenNullability<SomeData>(arg: sd, nArg: sd)
+    try assertEquals(actual: mixed.asNullable()?.num, expected: 33)
+    try assertEquals(actual: mixed.pAsNullable?.num, expected: 33)
+}
+
+// Swift ignores the variance and lets you force-cast to whatever you need, for better or worse.
+// This would *not* work with direct Swift interop.
+func testGenericVariance() throws {
+    let sd = SomeData(num: 22)
+
+    let variOut = GenVarOut<SomeData>(arg: sd)
+    let variOutAny : GenVarOut<BaseData> = variOut as! GenVarOut<BaseData>
+    let variOutOther : GenVarOut<SomeOtherData> = variOut as! GenVarOut<SomeOtherData>
+
+    let variOutCheck = "variOut: \(variOut.arg.asString()), variOutAny: \(variOutAny.arg.asString()), variOutOther: \(variOutOther.arg.asString())"
+    try assertEquals(actual: variOutCheck, expected: "variOut: 22, variOutAny: 22, variOutOther: 22")
+
+    let variIn = GenVarIn<SomeData>(tArg: sd)
+    let variInAny : GenVarIn<BaseData> = variIn as! GenVarIn<BaseData>
+    let variInOther : GenVarIn<SomeOtherData> = variIn as! GenVarIn<SomeOtherData>
+
+    let varInCheck = "variIn: \(variIn.valString()), variInAny: \(variInAny.valString()), variInOther: \(variInOther.valString())"
+    try assertEquals(actual: varInCheck, expected: "variIn: SomeData(num=22), variInAny: SomeData(num=22), variInOther: SomeData(num=22)")
+
+    let variCoType:GenVarOut<BaseData> = Values_genericsKt.variCoType()
+    try assertEquals(actual: "890", expected: variCoType.arg.asString())
+
+    let variContraType:GenVarIn<SomeData> = Values_genericsKt.variContraType()
+    try assertEquals(actual: "SomeData(num=1890)", expected: variContraType.valString())
+}
+
+// Swift should completely ignore this, as should objc. Really verifying that the header generator
+// deals with this
+func testGenericUseSiteVariance() throws {
+    let sd = SomeData(num: 22)
+
+    let varUse = GenVarUse<BaseData>(arg: sd)
+    let varUseArg = GenVarUse<BaseData>(arg: sd)
+
+    varUse.varUse(a: varUseArg, b: GenVarUse<SomeData>(arg: sd) as! GenVarUse<BaseData>)
+}
+
+func testGenericInterface() throws {
+    let a: NoGeneric = SomeGeneric<SomeData>(t: SomeData(num: 52))
+    try assertEquals(actual: (a.myVal() as! SomeData).num, expected: 52)
+}
+
+func testGenericInheritance() throws {
+    let ge = GenEx<SomeData, SomeOtherData>(myT:SomeOtherData(str:"Hello"), baseT:SomeData(num: 11))
+    try assertTrue(ge.t is SomeData, "base property not SomeData")
```

Comment:
AFAIK, this checks only the dynamic type, not the static one.
So this assert doesn't ensure that `ge.t` is statically known to be `SomeData` and can be assigned to a a `SomeData`-typed variable.

Reply:
Yeah, that check doesn't really do what I want. I've replaced how that works (in next push) for all type checks. Rather than check type, if the type isn't correct Swift needs a conversion, so the fact that it compiles and runs should imply that the types are as expected.

Reply:
What do you mean?

How does [the test](https://github.com/JetBrains/kotlin-native/blob/3f38a28068cd5951ab0c21e18181d9d1d2ebc2d1/backend.native/tests/framework/values_generics/values.swift#L150) now ensure that `ge.t` is declared as `SomeData`?

Reply:
That line is old. The [last push](https://github.com/kpgalligan/kotlin-native/blob/3f38a28068cd5951ab0c21e18181d9d1d2ebc2d1/backend.native/tests/framework/values_generics/values.swift#L150) was referencing the value instead of "is SomeData", which should fail to compile if not `SomeData`. 

```
try assertEquals(actual: ge.t.num, expected: 11)
```

However, I have some more testing to figure out as the properties and methods of the base class being written in the child class means disabling the superclass generics didn't cause that test to fail.

```
__attribute__((objc_subclassing_restricted))
__attribute__((swift_name("GenEx")))
@interface MainGenEx<TT, T> : MainGenBase
- (instancetype)initWithMyT:(T)myT baseT:(TT)baseT __attribute__((swift_name("init(myT:baseT:)"))) __attribute__((objc_designated_initializer));
- (instancetype)initWithT:(id)t __attribute__((swift_name("init(t:)"))) __attribute__((objc_designated_initializer)) __attribute__((unavailable));
- (TT)goBase __attribute__((swift_name("goBase()")));
@property (readonly) T myT __attribute__((swift_name("myT")));
@property (readonly) TT t __attribute__((swift_name("t")));
@end;
```

Property 't' is still SomeData even though we don't tell MainGenBase what `T` is because `t` is also written to `MainGenEx`. If we specify `T` as `Any`, Swift will report true with `is SomeData` but requires a cast to `SomeData`.

In [this example](https://github.com/kpgalligan/kotlin-native/blob/3f38a28068cd5951ab0c21e18181d9d1d2ebc2d1/backend.native/tests/framework/values_generics/values.swift#L157), GenExAny is defined as:

```
class GenExAny<TT:Any, T:Any>(val myT:T, baseT:TT):GenBase<Any>(baseT)
```

That forces the cast to `SomeData`. If GenEx was also writing `t` as `Any/id`, it should fail to compile.

However, I went in and disabled the generic superclass write to see what would fail and while one thing did fail, it was less than I expected. Will be diving into that more and update with better testing.

Reply:
`ge.t.num` doesn't fail because Swift seems to allow accessing any method or property on `id`.
That's why the proper way to test the type is to store the value into explicitly typed variable:
```
let geT: SomeData = ge.t
```

Reply:
Tests updated to specific types and assigned as above. I wasn't understanding why the parent class property was being written out, but I see that now. It [filters by string](https://github.com/JetBrains/kotlin-native/blob/ea11c377a61ea968c199fa6100a1bcbf7057f000/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportHeaderGenerator.kt#L392).

```objc
__attribute__((swift_name("GenEx")))
@interface MainGenEx<TT, T> : MainGenBase
- (instancetype)initWithMyT:(T)myT baseT:(TT)baseT __attribute__((swift_name("init(myT:baseT:)"))) __attribute__((objc_designated_initializer));
- (instancetype)initWithT:(id)t __attribute__((swift_name("init(t:)"))) __attribute__((objc_designated_initializer)) __attribute__((unavailable));
- (TT)goBase __attribute__((swift_name("goBase()")));
@property (readonly) T myT __attribute__((swift_name("myT")));
@property (readonly) TT t __attribute__((swift_name("t")));
@end;
```

I'm not sure why redefining the property exists in current code, so I'm a little reluctant to try to be smarter about filtering for generics. The possible parameter name differences make redefining more likely, but it should still work as expected.

I added an example that won't redefine the property to the test just to make sure that works.

```kotlin
open class GenBase<T:Any>(val t:T)
class GenEx<TT:Any, T:Any>(val myT:T, baseT:TT):GenBase<TT>(baseT)
class GenEx2<T:Any, S:Any>(val myT:S, baseT:T):GenBase<T>(baseT)
```

Can write a smarter filter, but want to be clear on the cases it exists for.

Reply:
> It filters by string.

This machinery has to be revised later. Let's keep current simple filter for now.